docs/015_How-to-use-Git-as-a-database.md

```diff
@@ -0,0 +1,423 @@
+# Using Git as a key-value database
+
+Obviously you can store whatever you want in a Git repository but in this article we will see how to use the Git internal data structures to store and retrieve data.
+
+- [Where to store things](#where-to-store-things)
+- [Using Git Objects](#using-git-objects)
+- [Using Git References](#using-git-references)
+- [Store only the latest state](#store-only-the-latest-state)
+- [Store state change history](#store-state-change-history)
+- [How to solve race conditions](#how-to-solve-race-conditions)
+  - [Optimistic concurrency control](#optimistic-concurrency-control)
+  - [Atomic Git push](#atomic-git-push)
+- [GitHub Git database](#github-git-database)
+- [Links](#links)
+  - [Projects using Git objects and references to store data](#projects-using-git-objects-and-references-to-store-data)
+  - [Talks](#talks)
+  - [Articles](#articles)
+  - [Books](#books)
+
+## Where to store things?
+
+With Git you can store data in two different ways:
+
+- [Git objects](https://git-scm.com/book/en/v2/Git-Internals-Git-Objects)
+- [References](https://git-scm.com/book/en/v2/Git-Internals-Git-References)
+
+Those are the two ways Git stores data internally.
+
+## Using Git Objects
+
+Git internally uses a key-value database with only 4 [types of objects](https://git-scm.com/book/en/v2/Git-Internals-Git-Objects): `blobs`, `trees`, `commits` and `annotated tags`.
+
+Each object is stored in the database and the way to reference the object is by using its [sha1](https://en.wikipedia.org/wiki/SHA-1) (a checksum of the content).
+
+For example, you can insert a new object in the database with:
+
+```s
+cd /tmp
+mkdir test
+cd test
+git init
+echo 'test content' | git hash-object -w --stdin
+d670460b4b4aece5915caf5c68d12f560a9fe3e4
+```
+
+The the last command output is the sha1 of the new object. You can get the object content back using that sha1:
```

Comment:
The "the" last command output is

Duplicated word

Reply:
Fixed.